File: src/test/java/msfol/SubstitutionTest.java
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import org.junit.Test;
import org.junit.BeforeClass;
import org.junit.Ignore;

import fortress.msfol.*;
import java.util.List;
import java.util.Set;
import java.util.ArrayList;
import java.io.*;
import fortress.util.Errors;

public class SubstitutionTest {
    
    private void assertAlphaEquivalent(Term expected, Term actual) {
        assertTrue(expected.toString() + "\n should be alpha equiv to the actual \n" + actual.toString(),
            expected.alphaEquivalent(actual));
    }
    
    Sort A = Sort.mkSortConst("A");
    Sort B = Sort.mkSortConst("B");
    
    Var w = Term.mkVar("w");
    Var x = Term.mkVar("x");
    Var y = Term.mkVar("y");
    Var z = Term.mkVar("z");
    Var a = Term.mkVar("a");
    Var b = Term.mkVar("b");
    Var c = Term.mkVar("c");
    Var p = Term.mkVar("p");
    Var q = Term.mkVar("q");
    
    Var _1 = Term.mkVar("_1");
    Var _2 = Term.mkVar("_2");
    
    @Test
    public void basicSubstitution() {
        Term t = Term.mkImp(p, Term.mkEq(x, Term.mkApp("f", y)));
        Term expected = Term.mkImp(p, Term.mkEq(Term.mkApp("h", z), Term.mkApp("f", y)));
        assertEquals(expected, t.substitute(x, Term.mkApp("h", z)));
    }
    
    @Test
    public void boundVar() {
        Term t1 = Term.mkAnd(Term.mkEq(x, x),
                       Term.mkExists(x.of(A), Term.mkApp("R", x)));
                       
        Term t2 = Term.mkAnd(Term.mkEq(x, x),
                       Term.mkForall(x.of(A), Term.mkApp("R", x)));
                       
        Term e1 = Term.mkAnd(Term.mkEq(Term.mkBottom(), Term.mkBottom()),
                      Term.mkExists(x.of(A), Term.mkApp("R", x)));
                      
        Term e2 = Term.mkAnd(Term.mkEq(Term.mkBottom(), Term.mkBottom()),
                      Term.mkForall(x.of(A), Term.mkApp("R", x)));
                      
        assertEquals(e1, t1.substitute(x, Term.mkBottom()));    
        assertEquals(e2, t2.substitute(x, Term.mkBottom()));    
    }
    
    @Test
    public void variableCaptureBasic() {
        Term t1 = Term.mkExists(x.of(A), Term.mkEq(x, y));
        Term t2 = Term.mkForall(x.of(A), Term.mkEq(x, y));
        
        Term e1 = Term.mkExists(z.of(A), Term.mkEq(z, x));
        Term e2 = Term.mkForall(z.of(A), Term.mkEq(z, x));
        
        Term r1 = t1.substitute(y, x);
        Term r2 = t2.substitute(y, x);
        assertAlphaEquivalent(e1, r1);
        assertAlphaEquivalent(e2, r2);
    }
    
    @Test
    public void variableCaptureMultivarQuantifier() {
        Term t1 = Term.mkExists(List.of(x.of(A), y.of(B)), Term.mkApp("P", x, y, z));
        Term t2 = Term.mkForall(List.of(x.of(A), y.of(B)), Term.mkApp("P", x, y, z));
        
        Term expected1 = Term.mkExists(List.of(x.of(A), w.of(B)), Term.mkApp("P", x, w, y));
        Term expected2 = Term.mkForall(List.of(w.of(A), y.of(B)), Term.mkApp("P", w, y, x));
        
        assertAlphaEquivalent(expected1, t1.substitute(z, y));
        assertAlphaEquivalent(expected2, t2.substitute(z, x));
    }
    
    @Test
    public void variableCaptureComplex() {
        Term t = Term.mkOr(
            Term.mkForall(x.of(A), Term.mkAnd(
                Term.mkApp("Q", x),
                Term.mkExists(x.of(A), Term.mkApp("P", x, y)))),
            Term.mkApp("Q", y));
        
        Term expected = Term.mkOr(
            Term.mkForall(z.of(A), Term.mkAnd(
                Term.mkApp("Q", z),
                Term.mkExists(w.of(A), Term.mkApp("P", w, x)))),
            Term.mkApp("Q", x));
        
        assertAlphaEquivalent(expected, t.substitute(y, x));
    }
    
}

File: src/test/scala/operations/SubstitutionTest.scala
import org.scalatest._
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner

import fortress.msfol._
import fortress.operations.TermOps._

@RunWith(classOf[JUnitRunner])
class SubstitutionTest extends FunSuite with Matchers {
    
    def assertAlphaEquivalent(actual: Term, expected: Term): Unit = {
        assert(actual alphaEquivalent expected,
            actual.toString + "\n is not alpha equivalent to \n" + expected.toString,
        )
    }
    
    val A = Sort.mkSortConst("A")
    val B = Sort.mkSortConst("B")
    
    val w = Term.mkVar("w")
    val x = Term.mkVar("x")
    val y = Term.mkVar("y")
    val z = Term.mkVar("z")
    val b = Term.mkVar("b")
    val c = Term.mkVar("c")
    val p = Term.mkVar("p")
    val q = Term.mkVar("q")
    
    val _1 = Term.mkVar("_1")
    val _2 = Term.mkVar("_2")
    
    test("basic substitution") {
        val t = p ==> (x === App("f", y))
        val expected = p ==> (App("h", z) === App("f", y))
        t.substitute(x, App("h", z)) should be (expected)
    }
    
    test("bound variable should not be substituted") {
        val t1 = And(x === x, Exists(x of A, App("R", x)))
                       
        val t2 = And(x === x, Forall(x of A, Term.mkApp("R", x)))
                       
        val e1 = And(Bottom === Bottom, Exists(x of A, App("R", x)))
                      
        val e2 = And(Bottom === Bottom, Forall(x of A, App("R", x)))
                      
        t1.substitute(x, Bottom) should be (e1)
        t2.substitute(x, Bottom) should be (e2)
    }
    
    test("variable capture basic") {
        val t1 = Exists(x of A, x === y)
        val t2 = Forall(x of A, x === y)
        
        val e1 = Exists(z of A, z === x)
        val e2 = Forall(z of A, z === x)
        
        val r1 = t1.substitute(y, x)
        val r2 = t2.substitute(y, x)
        assertAlphaEquivalent(r1, e1)
        assertAlphaEquivalent(r2, e2)
    }
    
    test("variable capture multivar quantifier") {
        val t1 = Exists(Seq(x of A, y of B), App("P", x, y, z))
        val t2 = Forall(Seq(x of A, y of B), App("P", x, y, z))
        
        val expected1 = Exists(Seq(x of A, w of B), App("P", x, w, y))
        val expected2 = Forall(Seq(w of A, y of B), App("P", w, y, x))
        
        assertAlphaEquivalent(t1.substitute(z, y), expected1)
        assertAlphaEquivalent(t2.substitute(z, x), expected2)
    }
    
    test("variable capture complex") {
        val t = Or(
            Forall(x of A, And(
                App("Q", x),
                Exists(x of A, App("P", x, y)))),
            App("Q", y))
        
        val expected = Or(
            Forall(z of A, And(
                App("Q", z),
                Exists(w of A, App("P", w, x)))),
            App("Q", x))
        
        assertAlphaEquivalent(t.substitute(y, x), expected);
    }
    
}
